File: Data Structures & Algorithms/rotting-fruit/submission-1.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        dr = [0, 0, -1, 1]
        dc = [1, -1, 0, 0]
        ret = 0
        while True:
            rottingSet = set()
            for r in range(len(grid)):
                for c in range(len(grid[0])):
                    if grid[r][c] != 1:
                        continue
                    
                    hasAdjacentFreshFruit = False
                    for i in range(4):
                        nr, nc = r + dr[i], c + dc[i]
                        if nc < 0 or nr < 0 or nc >= len(grid[0]) or nr >= len(grid):
                            continue
                        if grid[nr][nc] == 0:
                            continue
                        elif grid[nr][nc] == 2:
                            rottingSet.add((r, c))
                            hasAdjacentFreshFruit = True
                            break
                        elif grid[nr][nc] == 1:
                            hasAdjacentFreshFruit = True
                    
                    if not hasAdjacentFreshFruit:
                        return -1
            
            if len(rottingSet) == 0:
                break
            
            for r, c in rottingSet:
                grid[r][c] = 2

            ret += 1
        
        return ret
                    

File: Data Structures & Algorithms/swim-in-rising-water/submission-1.cpp
class Solution {
public:
    bool traverse(int i, int j, vector<vector<bool>> &visited, int max, const vector<vector<int>>& grid) {
        if (j < 0 || j == grid.size()) {
            return false;
        } else if (i < 0 || i == grid[0].size()) {
            return false;
        } else if (grid[j][i] > max) {
            return false;
        } else if (visited[j][i]) {
            return false;
        } else if (i == grid[0].size() - 1 && j == grid.size() - 1) return true;
        visited[j][i] = true;
        return traverse(i + 1, j, visited, max, grid) || traverse(i, j + 1, visited, max, grid) || traverse(i - 1, j, visited, max, grid) || traverse(i, j - 1, visited, max, grid);
    }
    int swimInWater(vector<vector<int>>& grid) {
        auto max{0};
        auto min{INT_MAX};
        for (const auto &g : grid) {
            for (const auto &c : g) {
                max = std::max(max, c);
                min = std::min(min, c);
            }
        }
        auto low{min};
        auto high{max};
        int mid{};
        auto ret{0};
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), 0));
        while (low < high) {
            mid = (low + high) / 2;
            auto temp{visited};
            if (traverse(0, 0, temp, mid, grid)) {
                high = mid;
            } else { 
                low = mid + 1;
            }
        }
        return high;
    }
};
